dotCMS/src/main/java/com/dotmarketing/portlets/workflows/actionlet/CommentOnWorkflowActionlet.java

```diff
@@ -1,58 +1,114 @@
+    /**
+     * 
+     */
+    private static final long serialVersionUID = 1L;
+    private static List<WorkflowActionletParameter> paramList = null;
+
+    @Override
+    public synchronized List<WorkflowActionletParameter> getParameters() {
+        if (paramList == null) {
+            paramList = new ArrayList<WorkflowActionletParameter>();
+            paramList.add(new WorkflowActionletParameter("comment", "Workflow Comment", null, true));
+        }
+        return paramList;
+    }
+
+    public String getName() {
+        return "Comment on Workflow";
+    }
+
+    public String getHowTo() {
+
+        return "This actionlet allows you to add a comment on the workflow task.";
+    }
+
+    public void executeAction(WorkflowProcessor processor, Map<String, WorkflowActionClassParameter> params)
+            throws WorkflowActionFailureException {
+        WorkflowActionClassParameter commentParam = params.get("comment");
+        WorkflowComment comment = new WorkflowComment();
+        final Contentlet con = processor.getContentlet();
+        Role role;
+        try {
+            role = APILocator.getRoleAPI().getUserRole(processor.getUser());
+            comment.setPostedBy(role.getId());
+        } catch (DotDataException e1) {
+            throw new WorkflowActionFailureException("unable to load role:" + processor.getUser(), e1);
+        }
+
 
-	@Override
-	public synchronized List<WorkflowActionletParameter> getParameters() {
-		if(paramList ==null){
-			paramList = new ArrayList<WorkflowActionletParameter>();
-			paramList.add(new WorkflowActionletParameter("comment", "Workflow Comment", null, true));
-		}
-		return paramList;
-	}
+        org.apache.velocity.context.Context ctx = VelocityUtil.getBasicContext();
+        ctx.put("workflow", processor);
+        ctx.put("user", processor.getUser());
+        ctx.put("contentlet", processor.getContentlet());
+        ctx.put("content", processor.getContentlet());
 
-	public String getName() {
-		return "Comment on Workflow";
-	}
+        try {
+            comment.setComment(VelocityUtil.eval(commentParam.getValue(), ctx));
+        } catch (Exception e1) {
+            Logger.warn(this.getClass(), "unable to parse comment, falling back" + e1);
+            comment.setComment(commentParam.getValue());
+        }
 
-	public String getHowTo() {
+        comment.setWorkflowtaskId(processor.getTask().getId());
 
-		return "This actionlet allows you to add a comment on the workflow task.";
-	}
+        if (processor.getTask().isNew()) {
+            try {
+                HibernateUtil.addAsyncCommitListener(() -> {
+                    try {
+                        WorkflowTask task = APILocator.getWorkflowAPI().findTaskByContentlet(con);
+                        comment.setWorkflowtaskId(task.getId());
+                        APILocator.getWorkflowAPI().saveComment(comment);
+                        DbConnectionFactory.closeAndCommit();
```

Comment:
I am not sure if this is needed, since the saveComment is transactional!

Reply:
Not sure about this either